### ISSUE TITLE
Suggestion: Introduce prettier

### ISSUE THREAD

**Issue by bd82** [jhipster/jhipster-core#158]
[Prettier](https://github.com/prettier/prettier) is an opinionated code formatter for JavaScript and assorted languages (json/css/typescript/...).

* Used in popular projects such as React/Babel/Webpack.
* Does formatting better than ESLint as it is a dedicated tool just for that.
* It allows for a consistent formatting between different projects and even different languages.
* Avoids arguments about irrelevant formatting styles as there are few options.
* Mainly helps completely forget about formatting as its auto fixing is very strong.

If you are interested I can create a PR that introduces prettier including everything required
for the development flows.

**Comment by MathieuAA** [comment]
I believe @jdubois and @deepu105 have (or had) a discussion in a ticket in the generator.

**Comment by bd82** [comment]
I will look for it

**Comment by bd82** [comment]
Found it: https://github.com/jhipster/generator-jhipster/issues/6379

That issue is about using prettier to format **generated** code from jhipster.
I'm suggesting it here for better(?) **development** flows of jhipster-core's **own** code base.

Up to you to decide if you like the concept of the prettier tool. 😄

**Comment by MathieuAA** [comment]
I'll think about it today!

**Comment by deepu105** [comment]
@bd82 if you add prettier don't you have to run some task to make it work on the code? either via eslint or pre-commit hooks? (I hate pre-commit hooks :smile: )
IMO `eslint --fix` already does a pretty good job for us but if you guys think it is really useful here then its ok.
If you decide to add it please make sure to align its options as closely as possible to the current style used. (Eslint and editorconfig)

**Comment by bd82** [comment]
Sorry one of my long posts here. 😄 

>  if you add prettier don't you have to run some task to make it work on the code?

Yes, but you usually do it via the editor integration options now a pre-commit hook.
-  https://github.com/prettier/prettier#editor-integration

> IMO eslint --fix already does a pretty good job for us

ESLint is not bad and would still remain for code quality rules.
-  https://github.com/prettier/prettier#how-does-it-compare-to-eslint-or-tslint-stylelint

The differences are:
-  In prettier **everything** is automatically fixable.
- Better implementation (working directly on the AST) so safer "fixing"
- Almost no options so no need to maintain a large configuration.
- Works on additional types of input (json/css/flow/...).
 
> If you decide to add it please make sure to align its options as closely as possible to the current style used. (Eslint and editorconfig)

This request highlights the **root paradigm difference**. 
If we decide to use prettier it means that we also decide that we no longer care about formatting.
It is a task that is **virtually** auto configured and with very few user definitions.

The editorConfig will be **removed** as it will become **redundant**.
And the eslintrc file will be **cut in half** as the syntax styles will be **removed**.

Basically it is a **subjective** question about **paradigms**:
- Would you **give up control** on trivial concerns like formatting
to get more time to spend on actually writing good?


The quality of the project is already good so the immediate gains will not be large.
But because I am willing to implement this and already done so in the past the cost would  be trivial too. So either choice is good 😄

**Comment by deepu105** [comment]
@bd82 yes I know about most of what you said. And I guess nobody spends time formatting these days anyway, at least I don't as I use IntelliJ and VScode primarily and both automatically format files based on my editorconfig and eslint settings. So for me this is just replacing one tool with another.
So, it would be nice to stick to current styling as much as possible(the options that are supported by prettier, like spacing, dangling coma etc) to be consistent across other JHipster projects as you said this is very subjective and unfortunately I'm one of those who prefer more control :smile_cat: 
So overall I'm ok with if @MathieuAA agrees.

**Comment by MathieuAA** [comment]
As Deepu said, as long as it's consistent with each and every JS JHipster project, I'm good.

**Comment by bd82** [comment]
> it's consistent with each and every JS JHipster project

I did not consider **other** projects in the jhipster org.
I think this will become too much work for too little gain.

**Comment by MathieuAA** [comment]
I don't agree, it can be done. Yes, it wouldn't take 5 minutes, but it's not impossible.

**Comment by bd82** [comment]
It is not impossible, just more than I was interested in investing.
I can implement it for this project in a PR and you can inspect if you are interested in applying the same to additional repos in this org?

**Comment by MathieuAA** [comment]
I don't think it's a priority

**Comment by deepu105** [comment]
@bd82 my opinions have changed about prettier in recent times after using it in some projects. It works really well and using it along with lint-staged and husky just works perfectly well.

@MathieuAA maybe we can reconsider this, if you are not against it

**Comment by bd82** [comment]
@MathieuAA should make the decision here as the main contributor and project lead
he will suffer the consequences in the long term (both good and bad).

I am willing to create the PR that integrates prettier with Jhipster-core if @MathieuAA wishes it.

**Comment by MathieuAA** [comment]
Let's consider, but later. I want to get v3.1 out ASAP. Le 5 juil. 2018 16:49, Deepu K Sasidharan <notifications@github.com> a écrit :@bd82 my opinions have changed about prettier in recent times after using it in some projects. It works really well and using it along with lint-staged and husky just works perfectly well.
@MathieuAA maybe we can reconsider this, if you are not against it

**Comment by MathieuAA** [comment]
Closing this, as a basic prettier support has just been added
